content/en/blog/_posts/2022-12-06-non-graceful-node-shutdown-to-beta.md

```diff
@@ -0,0 +1,78 @@
+
+## What is a Graceful Node Shutdown
+
+The [Graceful Node Shutdown](https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/) feature allows Kubelet to detect a node shutdown event, properly terminate the pods, and release resources, before the actual shutdown. Critical pods will be terminated after all the regular pods are terminated, to make sure the critical function of an application can work as long as possible.
+
+## What is a Non-Graceful Node Shutdown
+
+A Node Shutdown can be `Graceful` only if the node shutdown action can be detected by Kubelet’s Node Shutdown Manager. However, there are cases where a node shutdown action may not be detected by Kubelet's Node Shutdown Manager. This could happen either because the shutdown command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the `ShutdownGracePeriod` and `ShutdownGracePeriodCriticalPods` are not configured properly.
+
+When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, this becomes a non-graceful node shutdown. Non-graceful node shutdown is usually not a problem for stateless apps, however, it is a problem for stateful apps. If StatefulSet is used by the apps, the pod that is part of a StatefulSet will be stuck in `Terminating` status on the shutdown node and cannot move to a new running node. If Deployment is used by the apps, a new pod will come up on another running node but stuck in `ContainerCreating` status and the old pod will be stuck in `Terminating` status on the shutdown node. As a result, the application cannot function properly. If the original shutdown node comes up, the old pod will be deleted by Kubelet and the new pod will be created successfully on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.
+
+## What’s new in Beta
+
+With the promotion of the Non-Graceful Node Shutdown feature to beta, the feature gate  `NodeOutOfServiceVolumeDetach` is enabled by default on `kube-controller-manager` instead of being opt-in.
+
+New metrics `force_delete_pods_total` and `force_delete_pod_errors_total` are also added in the Pod GC Controller.
+
+## How does it work
+
+In the case of a node shutdown, if graceful shutdown is not working or node is in non-recoverable state due to hardware failure or broken OS, the user can manually add a `out-of-service` taint on the Node. For example, this can be `node.kubernetes.io/out-of-service=nodeshutdown:NoExecute` or `node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule`. This will trigger pods on the node to be forcefully deleted if there are no matching tolerations on the pods.  Persistent volumes attached to the shutdown node will be detached, and new pods will be created successfully on a different running node.
```

Comment:
```suggestion
In the case of a node shutdown, if a graceful shutdown is not working or the node is in a non-recoverable state due to hardware failure or broken OS, the user can manually add an `out-of-service` taint on the Node. For example, this can be `node.kubernetes.io/out-of-service=nodeshutdown:NoExecute` or `node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule`. This will trigger pods on the node to be forcefully deleted if there are no matching tolerations on the pods.  Persistent volumes attached to the shutdown node will be detached, and new pods will be created successfully on a different running node.
```

Reply:
Done